distributed/tests/test_worker.py

```diff
@@ -3161,85 +3161,84 @@ async def get_data(self, comm, *args, **kwargs):
     """
-    async with BlockedGetData(s.address) as a:
-        async with BlockedGatherDep(s.address) as b:
-            fut1 = c.submit(inc, 1, workers=[a.address], key="f1")
-            fut2 = c.submit(inc, fut1, workers=[a.address], key="f2")
-            await wait(fut2)
-            fut4 = c.submit(sum, fut1, fut2, workers=[b.address], key="f4")
-            fut3 = c.submit(inc, fut1, workers=[b.address], key="f3")
-
-            await _wait_for_state(fut2.key, b, "flight")
-            await b.in_gather_dep.wait()
-
-            fut4.release()
-            while fut4.key in b.tasks:
-                await asyncio.sleep(0)
+    async with BlockedGetData(s.address) as b:
+        fut1 = c.submit(inc, 1, workers=[b.address], key="f1")
+        await wait(fut1)
 
-            assert b.tasks[fut2.key].state == "cancelled"
+        fut2 = c.submit(inc, fut1, workers=[a.address], key="f2")
+        await b.in_get_data.wait()
+        assert a.tasks[fut1.key].state == "flight"
 
-            b.block_gather_dep.set()
-            await a.in_get_data.wait()
+        fut2.release()
+        while fut2.key in a.tasks:
+            await asyncio.sleep(0.01)
+        assert a.tasks[fut1.key].state == "cancelled"
 
-            fut4 = c.submit(sum, [fut1, fut2], workers=[b.address], key="f4")
-            await _wait_for_state(fut2.key, b, "flight")
+        fut2 = c.submit(inc, fut1, workers=[a.address], key="f2")
+        await _wait_for_state("f1", a, "flight")
 
-            a.block_get_data.set()
-            await wait([fut3, fut4])
-            f2_story = b.story(fut2.key)
-            assert f2_story
+        b.block_get_data.set()
+        assert await fut2 == 3
+        assert_story(
+            a.story(fut1.key),
+            [
+                (fut1.key, "fetch", "flight", "flight", {}),
+                (fut1.key, "flight", "released", "cancelled", {}),
+                (fut1.key, "cancelled", "fetch", "flight", {}),
+                (fut1.key, "flight", "memory", "memory", {"f2": "ready"}),
+            ],
+        )
+        # Test that the data transfer only happens once
+        assert_story(
+            a.story("request-dep"),
+            [
+                ("request-dep", b.address, {fut1.key}),
+            ],
+            strict=True,
+        )
 
 
 @gen_cluster(client=True, nthreads=[("", 1)])
 async def test_gather_dep_do_not_handle_response_of_not_requested_tasks(c, s, a):
-    """At time of writing, the gather_dep implementation filtered tasks again
-    for in-flight state. The response parser, however, did not distinguish
-    resulting in unwanted missing-data signals to the scheduler, causing
-    potential rescheduling or data leaks.
-    This test may become obsolete if the implementation changes significantly.
+    """A task transitions from flight to cancelled while gather_dep is waiting for the
+    data.
+
+    See also test_gather_dep_cancelled_rescheduled
```

Comment:
Remove assumption that both dependencies of a task will be gathered in a single fetch (this will no longer be the case in #6462).

Reply:
Where do we assume this in the previous test?

Reply:
same as above

Reply:
It does not.
50% of the times you get:
```
>           assert b.tasks[fut2.key].state == "flight"
E           AssertionError: assert 'fetch' == 'flight'
```